File: PInvoke/Parser/ExpressionParser.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.
Imports System.Globalization
Imports System.Text.RegularExpressions
Imports PInvoke.Contract

Namespace Parser

    ''' <summary>
    ''' Kind of expression
    ''' </summary>
    ''' <remarks></remarks>
    Public Enum ExpressionKind

        ' Binary operation such as +,-,/ 
        ' Token: Operation
        BinaryOperation

        ' '-' operation.  Left is the value
        NegativeOperation

        ' ! operation, Left is the value
        NegationOperation

        ' Token is the name of the function.  
        ' Left: Value
        ' Right: , if there are more arguments
        FunctionCall

        List

        ' Token: Target Type
        ' Left: Source that is being cast
        Cast

        ' Token: Value of the expression
        Leaf
    End Enum

    ''' <summary>
    ''' Expression Node
    ''' </summary>
    ''' <remarks></remarks>
    <DebuggerDisplay("{DisplayString}")> _
    Public Class ExpressionNode
        Private m_kind As ExpressionKind
        Private m_left As ExpressionNode
        Private m_right As ExpressionNode
        Private m_token As Token
        Private m_parenthesized As Boolean
        Private m_tag As Object

        Public Property Kind() As ExpressionKind
            Get
                Return m_kind
            End Get
            Set(ByVal value As ExpressionKind)
                m_kind = value
            End Set
        End Property

        Public Property LeftNode() As ExpressionNode
            Get
                Return m_left
            End Get
            Set(ByVal value As ExpressionNode)
                m_left = value
            End Set
        End Property

        Public Property RightNode() As ExpressionNode
            Get
                Return m_right
            End Get
            Set(ByVal value As ExpressionNode)
                m_right = value
            End Set
        End Property

        Public Property Token() As Token
            Get
                Return m_token
            End Get
            Set(ByVal value As Token)
                m_token = value
            End Set
        End Property

        Public Property Parenthesized() As Boolean
            Get
                Return m_parenthesized
            End Get
            Set(ByVal value As Boolean)
                m_parenthesized = True
            End Set
        End Property

        Public Property Tag() As Object
            Get
                Return m_tag
            End Get
            Set(ByVal value As Object)
                m_tag = value
            End Set
        End Property

        Public ReadOnly Property DisplayString() As String
            Get
                Dim str As String = String.Empty
                If m_left IsNot Nothing Then
                    str &= "(Left: " & m_left.DisplayString & ")"
                End If

                If m_right IsNot Nothing Then
                    str &= "(Right: " & m_right.DisplayString & ")"
                End If

                If Not String.IsNullOrEmpty(str) Then
                    str = " " & str
                End If

                If m_token Is Nothing Then
                    Return "Nothing" & str
                Else
                    Return m_token.Value & str
                End If
            End Get
        End Property

        Public Sub New(ByVal kind As ExpressionKind, ByVal value As Token)
            m_kind = kind
            m_token = value
        End Sub

        Public Shared Function CreateLeaf(ByVal bValue As Boolean) As ExpressionNode
            Dim token As Token
            If bValue Then
                token = New Token(TokenType.TrueKeyword, "true")
            Else
                token = New Token(TokenType.TrueKeyword, "false")
            End If

            Return New ExpressionNode(ExpressionKind.Leaf, token)
        End Function

        Public Shared Function CreateLeaf(ByVal number As Integer) As ExpressionNode
            Return New ExpressionNode(ExpressionKind.Leaf, New Token(TokenType.Number, number.ToString()))
        End Function

    End Class

    ''' <summary>
    ''' Converts an expression into an expression tree
    ''' </summary>
    ''' <remarks></remarks>
    Public Class ExpressionParser

        Public Function Parse(ByVal expression As String) As ExpressionNode
            Dim node As ExpressionNode = Nothing
            If Not Me.TryParse(expression, node) Then
                Throw New InvalidOperationException("Unable to parse the expression")
            End If

            Return node
        End Function

        Public Function IsParsable(ByVal expression As String) As Boolean
            Dim node As ExpressionNode = Nothing
            Return TryParse(expression, node)
        End Function

        Public Function TryParse(ByVal tokens As List(Of Token), ByRef node As ExpressionNode) As Boolean
            If tokens Is Nothing Then : Throw New ArgumentNullException("tokens") : End If

            Return TryParseComplete(tokens, node)
        End Function

        Public Function TryParse(ByVal expression As String, ByRef node As ExpressionNode) As Boolean
            If expression Is Nothing Then : Throw New ArgumentNullException("expression") : End If

            Using reader As New IO.StringReader(expression)
                Dim scanner As New Scanner(reader)
                scanner.Options.HideNewLines = True
                scanner.Options.HideComments = True
                scanner.Options.HideWhitespace = True
                scanner.Options.ThrowOnEndOfStream = False
                Return TryParseComplete(scanner.Tokenize(), node)
            End Using
        End Function

        Private Function TryParseComplete(ByVal tokens As List(Of Token), ByRef node As ExpressionNode) As Boolean
            Dim cur As ExpressionNode = Nothing
            Dim remaining As List(Of Token) = Nothing
            If Not TryParseCore(tokens, cur, remaining) Then
                Return False
            End If

            If remaining.Count = 0 Then
                node = cur
                Return True
            ElseIf Not remaining(0).IsBinaryOperation _
                AndAlso cur.Parenthesized _
                AndAlso cur.Kind = ExpressionKind.Leaf _
                AndAlso cur.Token.IsAnyWord Then

                ' This is a cast
                cur.Kind = ExpressionKind.Cast
                node = cur
                Return TryParseComplete(remaining, node.LeftNode)
            ElseIf remaining.Count = 1 OrElse Not remaining(0).IsBinaryOperation Then
                Return False
            Else
                Dim right As ExpressionNode = Nothing
                If Not TryParseComplete(remaining.GetRange(1, remaining.Count - 1), right) Then
                    Return False
                End If

                node = New ExpressionNode(ExpressionKind.BinaryOperation, remaining(0))
                node.LeftNode = cur
                node.RightNode = right
                Return True
            End If
        End Function

        Private Function TryParseCore(ByVal tokens As List(Of Token), ByRef node As ExpressionNode, ByRef remaining As List(Of Token)) As Boolean
            ThrowIfNull(tokens)

            If tokens.Count = 0 Then
                Return False
            End If

            ' Single tokens are the easiest
            If tokens.Count = 1 Then
                remaining = New List(Of Token)
                Return TryConvertTokenToExpressionLeafNode(tokens(0), node)
            End If

            Dim leftNode As ExpressionNode = Nothing
            Dim unaryNode As ExpressionNode = Nothing
            Dim nextIndex As Integer = -1

            If tokens.Count > 2 _
                AndAlso tokens(0).IsAnyWord _
                AndAlso tokens(1).TokenType = TokenType.ParenOpen Then

                ' Function call
                Return TryParseFunctionCall(tokens, node, remaining)
            ElseIf tokens(0).TokenType = TokenType.Bang Then
                node = New ExpressionNode(ExpressionKind.NegationOperation, tokens(0))
                Return TryParseCore(tokens.GetRange(1, tokens.Count - 1), node.LeftNode, remaining)
            ElseIf tokens(0).TokenType = TokenType.OpMinus Then
                node = New ExpressionNode(ExpressionKind.NegativeOperation, tokens(0))
                Return TryParseCore(tokens.GetRange(1, tokens.Count - 1), node.LeftNode, remaining)
            ElseIf tokens(0).TokenType = TokenType.ParenOpen Then
                Return TryParseParenExpression(tokens, node, remaining)
            ElseIf tokens.Count > 2 Then
                ' Has to be an operation so convert the left node to a leaf expression
                remaining = tokens.GetRange(1, tokens.Count - 1)
                Return TryConvertTokenToExpressionLeafNode(tokens(0), node)
            Else
                Return False
            End If
        End Function

        Private Function TryParseFunctionCall(ByVal tokens As List(Of Token), ByRef node As ExpressionNode, ByRef remaining As List(Of Token)) As Boolean
            ThrowIfTrue(tokens.Count < 3)

            node = New ExpressionNode(ExpressionKind.FunctionCall, tokens(0))

            ' Find the last index 
            Dim endIndex As Integer = FindMatchingParenIndex(tokens, 2)
            If endIndex = -1 Then
                Return False
            End If

            ' If there is more than just word() then there are arguments
            If tokens.Count > 3 Then
                Dim subList As List(Of Token) = tokens.GetRange(2, endIndex - 2)
                If Not TryParseFunctionCallArguments(subList, node) Then
                    Return False
                End If
            End If

            remaining = tokens.GetRange(endIndex + 1, tokens.Count - (endIndex + 1))
            Return True
        End Function

        Private Function TryParseFunctionCallArguments(ByVal tokens As List(Of Token), ByVal callNode As ExpressionNode) As Boolean
            ThrowIfNull(callNode)
            ThrowIfFalse(ExpressionKind.FunctionCall = callNode.Kind)

            ' Start the list
            Dim cur As ExpressionNode = callNode

            While tokens.Count > 0
                Dim index As Integer = FindNextCallArgumentSeparator(tokens)
                If index < 0 Then
                    ' No more separators so just parse out the rest of the tokens as an argument
                    If Not TryParseComplete(tokens, cur.LeftNode) Then
                        Return False
                    End If
                    tokens.Clear()
                Else
                    If Not TryParseComplete(tokens.GetRange(0, index), cur.LeftNode) Then
                        Return False
                    End If

                    tokens = tokens.GetRange(index + 1, tokens.Count - (index + 1))
                End If

                If tokens.Count > 0 Then
                    cur.RightNode = New ExpressionNode(ExpressionKind.List, New Token(TokenType.Comma, ","))
                    cur = cur.RightNode
                End If
            End While

            Return True
        End Function

        Private Function TryParseParenExpression(ByVal tokens As List(Of Token), ByRef node As ExpressionNode, ByRef remaining As List(Of Token)) As Boolean
            Dim endIndex As Integer = FindMatchingParenIndex(tokens, 1)
            If endIndex = -1 Then
                Return False
            End If

            remaining = tokens.GetRange(endIndex + 1, tokens.Count - (endIndex + 1))
            Dim success As Boolean = TryParseComplete(tokens.GetRange(1, endIndex - 1), node)
            If success Then
                node.Parenthesized = True
            End If

            Return success
        End Function

        Private Function FindNextCallArgumentSeparator(ByVal tokens As List(Of Token)) As Integer
            ThrowIfNull(tokens)

            For i As Integer = 0 To tokens.Count - 1
                If tokens(i).TokenType = TokenType.Comma Then
                    Return i
                End If
            Next

            Return -1
        End Function

        Private Function TryConvertTokenToExpressionBinaryOperation(ByVal token As Token, ByRef node As ExpressionNode) As Boolean
            ThrowIfNull(token)

            Dim isvalid As Boolean = token.IsBinaryOperation
            If isvalid Then
                node = New ExpressionNode(ExpressionKind.BinaryOperation, token)
                Return True
            Else
                node = Nothing
                Return False
            End If
        End Function

        Private Function TryConvertTokenToExpressionLeafNode(ByVal token As Token, ByRef node As ExpressionNode) As Boolean
            ThrowIfNull(token)

            If token.IsAnyWord Then
                node = New ExpressionNode(ExpressionKind.Leaf, token)
            ElseIf token.IsNumber Then
                node = New ExpressionNode(ExpressionKind.Leaf, token)
            ElseIf token.IsQuotedString Then
                node = New ExpressionNode(ExpressionKind.Leaf, token)
            ElseIf token.IsCharacter Then
                node = New ExpressionNode(ExpressionKind.Leaf, token)
            ElseIf token.TokenType = TokenType.TrueKeyword OrElse token.TokenType = TokenType.FalseKeyword Then
                node = New ExpressionNode(ExpressionKind.Leaf, token)
            Else
                node = Nothing
                Return False
            End If

            Return True
        End Function

        Private Function FindMatchingParenIndex(ByVal tokens As List(Of Token), ByVal start As Integer) As Integer

            Dim depth As Integer = 1
            For i As Integer = start To tokens.Count - 1
                Select Case tokens(i).TokenType
                    Case TokenType.ParenOpen
                        depth += 1
                    Case TokenType.ParenClose
                        depth -= 1
                        If 0 = depth Then
                            Return i
                        End If
                End Select
            Next

            Return -1
        End Function

    End Class

End Namespace

File: PInvoke.Core/Parser/ExpressionParser.cs
// Copyright (c) Microsoft Corporation.  All rights reserved.

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Text.RegularExpressions;
using static PInvoke.Contract;

namespace PInvoke.Parser
{

    /// <summary>
    /// Kind of expression
    /// </summary>
    /// <remarks></remarks>
    public enum ExpressionKind
    {

        // Binary operation such as +,-,/ 
        // Token: Operation
        BinaryOperation,

        // '-' operation.  Left is the value
        NegativeOperation,

        // ! operation, Left is the value
        NegationOperation,

        // Token is the name of the function.  
        // Left: Value
        // Right: , if there are more arguments
        FunctionCall,

        List,

        // Token: Target Type
        // Left: Source that is being cast
        Cast,

        // Token: Value of the expression
        Leaf
    }

    /// <summary>
    /// Expression Node
    /// </summary>
    /// <remarks></remarks>
    [DebuggerDisplay("{DisplayString}")]
    public class ExpressionNode
    {
        private ExpressionKind _kind;
        private Token _token;
        private bool _parenthesized;

        private object _tag;
        public ExpressionKind Kind
        {
            get { return _kind; }
            set { _kind = value; }
        }

        public ExpressionNode LeftNode;

        public ExpressionNode RightNode;

        public Token Token
        {
            get { return _token; }
            set { _token = value; }
        }

        public bool Parenthesized
        {
            get { return _parenthesized; }
            set { _parenthesized = value; }
        }

        // CTODO: Delete this.  It's a hack
        public object Tag
        {
            get { return _tag; }
            set { _tag = value; }
        }

        public string DisplayString
        {
            get
            {
                string str = string.Empty;
                if (LeftNode != null)
                {
                    str += "(Left: " + LeftNode.DisplayString + ")";
                }

                if (RightNode != null)
                {
                    str += "(Right: " + RightNode.DisplayString + ")";
                }

                if (!string.IsNullOrEmpty(str))
                {
                    str = " " + str;
                }

                if (_token == null)
                {
                    return "Nothing" + str;
                }
                else
                {
                    return _token.Value + str;
                }
            }
        }

        public ExpressionNode(ExpressionKind kind, Token value)
        {
            _kind = kind;
            _token = value;
        }

        public static ExpressionNode CreateLeaf(bool bValue)
        {
            Token token = default(Token);
            if (bValue)
            {
                token = new Token(TokenType.TrueKeyword, "true");
            }
            else
            {
                token = new Token(TokenType.TrueKeyword, "false");
            }

            return new ExpressionNode(ExpressionKind.Leaf, token);
        }

        public static ExpressionNode CreateLeaf(int number)
        {
            return new ExpressionNode(ExpressionKind.Leaf, new Token(TokenType.Number, number.ToString()));
        }

    }

    /// <summary>
    /// Converts an expression into an expression tree
    /// </summary>
    /// <remarks></remarks>
    public class ExpressionParser
    {

        public ExpressionNode Parse(string expression)
        {
            ExpressionNode node = null;
            if (!this.TryParse(expression, out node))
            {
                throw new InvalidOperationException("Unable to parse the expression");
            }

            return node;
        }

        public bool IsParsable(string expression)
        {
            ExpressionNode node = null;
            return TryParse(expression, out node);
        }

        public bool TryParse(List<Token> tokens, out ExpressionNode node)
        {
            if (tokens == null)
            {
                throw new ArgumentNullException("tokens");
            }

            return TryParseComplete(tokens, out node);
        }

        public bool TryParse(string expression, out ExpressionNode node)
        {
            if (expression == null)
            {
                throw new ArgumentNullException("expression");
            }

            using (var reader = new StringReader(expression))
            {
                Scanner scanner = new Scanner(reader);
                scanner.Options.HideNewLines = true;
                scanner.Options.HideComments = true;
                scanner.Options.HideWhitespace = true;
                scanner.Options.ThrowOnEndOfStream = false;
                return TryParseComplete(scanner.Tokenize(), out node);
            }
        }

        private bool TryParseComplete(List<Token> tokens, out ExpressionNode node)
        {
            ExpressionNode cur = null;
            List<Token> remaining = null;
            if (!TryParseCore(tokens, out cur, out remaining))
            {
                node = null;
                return false;
            }

            if (remaining.Count == 0)
            {
                node = cur;
                return true;

            }
            else if (!remaining[0].IsBinaryOperation && cur.Parenthesized && cur.Kind == ExpressionKind.Leaf && cur.Token.IsAnyWord)
            {
                // This is a cast
                cur.Kind = ExpressionKind.Cast;
                node = cur;
                return TryParseComplete(remaining, out node.LeftNode);
            }
            else if (remaining.Count == 1 || !remaining[0].IsBinaryOperation)
            {
                node = null;
                return false;
            }
            else
            {
                ExpressionNode right = null;
                if (!TryParseComplete(remaining.GetRange(1, remaining.Count - 1), out right))
                {
                    node = null;
                    return false;
                }

                node = new ExpressionNode(ExpressionKind.BinaryOperation, remaining[0]);
                node.LeftNode = cur;
                node.RightNode = right;
                return true;
            }
        }

        private bool TryParseCore(List<Token> tokens, out ExpressionNode node, out List<Token> remaining)
        {
            ThrowIfNull(tokens);

            if (tokens.Count == 0)
            {
                node = null;
                remaining = null;
                return false;
            }

            // Single tokens are the easiest
            if (tokens.Count == 1)
            {
                remaining = new List<Token>();
                return TryConvertTokenToExpressionLeafNode(tokens[0], out node);
            }

            if (tokens.Count > 2 && tokens[0].IsAnyWord && tokens[1].TokenType == TokenType.ParenOpen)
            {
                // Function call
                return TryParseFunctionCall(tokens, out node, out remaining);
            }
            else if (tokens[0].TokenType == TokenType.Bang)
            {
                node = new ExpressionNode(ExpressionKind.NegationOperation, tokens[0]);
                return TryParseCore(tokens.GetRange(1, tokens.Count - 1), out node.LeftNode, out remaining);
            }
            else if (tokens[0].TokenType == TokenType.OpMinus)
            {
                node = new ExpressionNode(ExpressionKind.NegativeOperation, tokens[0]);
                return TryParseCore(tokens.GetRange(1, tokens.Count - 1), out node.LeftNode, out remaining);
            }
            else if (tokens[0].TokenType == TokenType.ParenOpen)
            {
                return TryParseParenExpression(tokens, out node, out remaining);
            }
            else if (tokens.Count > 2)
            {
                // Has to be an operation so convert the left node to a leaf expression
                remaining = tokens.GetRange(1, tokens.Count - 1);
                return TryConvertTokenToExpressionLeafNode(tokens[0], out node);
            }
            else
            {
                node = null;
                remaining = null;
                return false;
            }
        }

        private bool TryParseFunctionCall(List<Token> tokens, out ExpressionNode node, out List<Token> remaining)
        {
            ThrowIfTrue(tokens.Count < 3);

            node = new ExpressionNode(ExpressionKind.FunctionCall, tokens[0]);

            // Find the last index 
            int endIndex = FindMatchingParenIndex(tokens, 2);
            if (endIndex == -1)
            {
                remaining = null;
                return false;
            }

            // If there is more than just word() then there are arguments
            if (tokens.Count > 3)
            {
                List<Token> subList = tokens.GetRange(2, endIndex - 2);
                if (!TryParseFunctionCallArguments(subList, node))
                {
                    remaining = null;
                    return false;
                }
            }

            remaining = tokens.GetRange(endIndex + 1, tokens.Count - (endIndex + 1));
            return true;
        }

        private bool TryParseFunctionCallArguments(List<Token> tokens, ExpressionNode callNode)
        {
            ThrowIfNull(callNode);
            ThrowIfFalse(ExpressionKind.FunctionCall == callNode.Kind);

            // Start the list
            ExpressionNode cur = callNode;

            while (tokens.Count > 0)
            {
                int index = FindNextCallArgumentSeparator(tokens);
                if (index < 0)
                {
                    // No more separators so just parse out the rest of the tokens as an argument
                    if (!TryParseComplete(tokens, out cur.LeftNode))
                    {
                        return false;
                    }
                    tokens.Clear();
                }
                else
                {
                    if (!TryParseComplete(tokens.GetRange(0, index), out cur.LeftNode))
                    {
                        return false;
                    }

                    tokens = tokens.GetRange(index + 1, tokens.Count - (index + 1));
                }

                if (tokens.Count > 0)
                {
                    cur.RightNode = new ExpressionNode(ExpressionKind.List, new Token(TokenType.Comma, ","));
                    cur = cur.RightNode;
                }
            }

            return true;
        }

        private bool TryParseParenExpression(List<Token> tokens, out ExpressionNode node, out List<Token> remaining)
        {
            node = null;
            remaining = null;

            int endIndex = FindMatchingParenIndex(tokens, 1);
            if (endIndex == -1)
            {
                return false;
            }

            remaining = tokens.GetRange(endIndex + 1, tokens.Count - (endIndex + 1));
            bool success = TryParseComplete(tokens.GetRange(1, endIndex - 1), out node);
            if (success)
            {
                node.Parenthesized = true;
            }

            return success;
        }

        private int FindNextCallArgumentSeparator(List<Token> tokens)
        {
            ThrowIfNull(tokens);

            for (int i = 0; i <= tokens.Count - 1; i++)
            {
                if (tokens[i].TokenType == TokenType.Comma)
                {
                    return i;
                }
            }

            return -1;
        }

        private bool TryConvertTokenToExpressionBinaryOperation(Token token, out ExpressionNode node)
        {
            ThrowIfNull(token);

            bool isvalid = token.IsBinaryOperation;
            if (isvalid)
            {
                node = new ExpressionNode(ExpressionKind.BinaryOperation, token);
                return true;
            }
            else
            {
                node = null;
                return false;
            }
        }

        private bool TryConvertTokenToExpressionLeafNode(Token token, out ExpressionNode node)
        {
            ThrowIfNull(token);

            if (token.IsAnyWord)
            {
                node = new ExpressionNode(ExpressionKind.Leaf, token);
            }
            else if (token.IsNumber)
            {
                node = new ExpressionNode(ExpressionKind.Leaf, token);
            }
            else if (token.IsQuotedString)
            {
                node = new ExpressionNode(ExpressionKind.Leaf, token);
            }
            else if (token.IsCharacter)
            {
                node = new ExpressionNode(ExpressionKind.Leaf, token);
            }
            else if (token.TokenType == TokenType.TrueKeyword || token.TokenType == TokenType.FalseKeyword)
            {
                node = new ExpressionNode(ExpressionKind.Leaf, token);
            }
            else
            {
                node = null;
                return false;
            }

            return true;
        }

        private int FindMatchingParenIndex(List<Token> tokens, int start)
        {

            int depth = 1;
            for (int i = start; i <= tokens.Count - 1; i++)
            {
                switch (tokens[i].TokenType)
                {
                    case TokenType.ParenOpen:
                        depth += 1;
                        break;
                    case TokenType.ParenClose:
                        depth -= 1;
                        if (0 == depth)
                        {
                            return i;
                        }
                        break;
                }
            }

            return -1;
        }

    }

}
